twister: add a --flash-command flag

fabiobaltieri opened the pull request:
Currently twister requires west to flash a board, either directly using west-flash or indirectly through the cmake flash target.

Add an option to specify a custom flash script, this is passed a build-dir and board-id flags so it can be used to implement custom flashing scripts in a system that does not use west.

fabiobaltieri committed:
twister: add a --flash-command flag

Currently twister requires west to flash a board, either directly using
west-flash or indirectly through the cmake flash target.

Add an option to specify a custom flash script, this is passed a
build-dir and board-id flags so it can be used to implement custom
flashing scripts in a system that does not use west.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

marc-hb commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -840,6 +840,15 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
     )
+    parser.add_argument(
+        "--flash-command",
+        help="""Uses a custom flash command instead of ninja or make to flash
```
Did you mean?
```suggestion
        help="""Uses a custom flash command instead of west to flash
```
Context:
- #92406

marc-hb replied:
At least @tejlmand seems to care; in #92406 (which this supersedes - so it was IMHO worth linking them to each other) he just said `ninja flash` should die. So it did not seem like a good idea to mention that some `--help` text.

> dropped that whole "instead of bit"

Great!

fabiobaltieri commented on doc/develop/test/twister.rst:
```diff
@@ -1443,6 +1443,11 @@ work. It is equivalent to following west and twister commands.
+``flash-command`` flag. The script is called with a ``--build-dir`` with the
+path of the current build, as well as a ``--board-id`` flag to identify the
+specific device when multiple are available in a hardware map
+
```
sure why not, the use case I think it's pretty clear already? "Where west is not available", feel free to suggest a better wording

nashif commented:
we should first look into implementing this at the cmake level instead of making this yet another twister option. Right now, 'ninja flash' is either using `west` or erroring and aksing you to create a west workspace, `ninja flash` in the absense of WEST and based on some definition in the board cmake should be able to run any custom scripts and skip west completely.

something like this, quickly hacked:

```
diff --git a/boards/nxp/frdm_k64f/board.cmake b/boards/nxp/frdm_k64f/board.cmake
index d2fafeb4a00..e732a1e9bbc 100644
--- a/boards/nxp/frdm_k64f/board.cmake
+++ b/boards/nxp/frdm_k64f/board.cmake
@@ -1,5 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0

+set(CUSTOM_FLASHER "/tmp/blah.sh")
 board_runner_args(jlink "--device=MK64FN1M0xxx12")
 board_runner_args(linkserver  "--device=MK64FN1M0xxx12:FRDM-K64F")

diff --git a/cmake/flash/CMakeLists.txt b/cmake/flash/CMakeLists.txt
index 985b7536154..5c609428410 100644
--- a/cmake/flash/CMakeLists.txt
+++ b/cmake/flash/CMakeLists.txt
@@ -186,7 +186,27 @@ foreach(target flash debug debugserver attach rtt)
     set(RUNNER_VERBOSE)
   endif()

-  if(WEST)
+  if(CUSTOM_FLASHER)
+    add_custom_target(${target}
+      # This script will print an error message and fail if <target> has added
+      # dependencies. This is done using dedicated CMake script, as
+      # `cmake -E {true|false}` is not available until CMake 3.16.
+      COMMAND ${CMAKE_COMMAND}
+        -DTARGET=${target}
+        -DDEPENDENCIES="$<TARGET_PROPERTY:${target},MANUALLY_ADDED_DEPENDENCIES>"
+        -P ${CMAKE_CURRENT_LIST_DIR}/check_runner_dependencies.cmake
+      COMMAND
+        ${CMAKE_COMMAND} -E env ZEPHYR_BASE=${ZEPHYR_BASE}
+       ${CUSTOM_FLASHER}
+        ${RUNNER_VERBOSE}
+        ${target}
+      WORKING_DIRECTORY
+        ${APPLICATION_BINARY_DIR}
+      COMMENT
+        ${comment}
+      USES_TERMINAL
+    )
+  elseif(WEST)
     add_custom_target(${target}
       # This script will print an error message and fail if <target> has added
       # dependencies. This is done using dedicated CMake script, as
@@ -212,5 +232,5 @@ foreach(target flash debug debugserver attach rtt)
           '${CMAKE_MAKE_PROGRAM} ${target}', please create a west workspace.\"
       USES_TERMINAL
       )
-  endif(WEST)
+  endif()
 endforeach()
```

with /tmp/blah.sh:
```
#!/bin/sh
#
echo "BLAH"
```


then run `ninja flash` and ...

```
...
device_CMSIS component is included.
-- Using ccache: /usr/bin/ccache
-- Found gen_kobject_list: /home/nashif/zephyrproject/zephyr/scripts/build/gen_kobject_list.py
-- Configuring done (5.2s)
-- Generating done (0.2s)
-- Build files have been written to: /home/nashif/zephyrproject/zephyr/build
[0/1] Flashing frdm_k64f
BLAH
```

fabiobaltieri committed:
twister: pass flash_command to the harness

Add the necessary logic to pass the custom flash_command to the pytest
harness.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

gchwier reviewed (commented):
I was tested with that script:
```python
#!/usr/bin/env python3

"""
Custom flash script for Twister device testing.
This script takes --build-dir and --board-id parameters and calls west flash.
"""

import argparse
import subprocess
import sys


def main():
    parser = argparse.ArgumentParser(description='Custom flash script for Twister')
    parser.add_argument('--build-dir', required=True,
                        help='Path to the build directory')
    parser.add_argument('--board-id',
                        help='Board ID to identify the specific device')

    args = parser.parse_args()

    # Construct the west flash command
    west_cmd = [
        'west', 'flash',
        '--skip-rebuild',
        '-d', args.build_dir
    ]
    if args.board_id:
        west_cmd.extend(['--dev-id', args.board_id])

    print(f"Executing: {' '.join(west_cmd)}")

    try:
        # Execute the west flash command
        result = subprocess.run(west_cmd, check=True, capture_output=True, text=True)
        print("Flash operation completed successfully")
        print(result.stdout)
        return 0
    except subprocess.CalledProcessError as e:
        print(f"Flash operation failed with return code {e.returncode}")
        print(f"stdout: {e.stdout}")
        print(f"stderr: {e.stderr}")
        return e.returncode
    except Exception as e:
        print(f"Unexpected error: {e}")
        return 1


if __name__ == '__main__':
    sys.exit(main())
```

with commands:
`./scripts/twister -vv -ll debug -T samples/hello_world --device-testing -p nrf52840dk/nrf52840 --device-serial /dev/ttyACM0 --no-clean --flash-command ./custom_flash_script.py`
`./scripts/twister -vv -ll debug -T samples/subsys/testsuite/pytest/shell -s sample.pytest.shell --device-testing -p nrf52840dk/nrf52840 --device-serial /dev/ttyACM0 --no-clean --flash-command=./custom_flash_script.py`
and with hardware map (then dev id is also provided)

fabiobaltieri commented on doc/develop/test/twister.rst:
```diff
@@ -1443,6 +1443,29 @@ work. It is equivalent to following west and twister commands.
+      .. code-block:: bash
+
+         twister -p npcx9m6f_evb --device-testing --device-serial /dev/ttyACM0
+         --flash-command custom_flash_script.py,--flag1,--flag2
```
yup fair enough, updated

gchwier commented on doc/develop/test/twister.rst:
```diff
@@ -1443,6 +1443,29 @@ work. It is equivalent to following west and twister commands.
+
+         python .\scripts\twister -p npcx9m6f_evb --device-testing
+         --device-serial COM1
+         --flash-command ./custom_flash_script.py,--flag1,--flag2
```
I'm not sure if `./` works on windows

fabiobaltieri replied:
Oh, right, Windows, that's what this line is, dropped.

fabiobaltieri commented:
Tested on my setup as well, something like `./twister -n -v -p npcx9/npcx9m7f --toolchain=coreboot-sdk --device-testing --hardware-map hwmap --flash-command ../dagwood/flash.py --testsuite-root=../../../src/third_party/zephyr/main/samples -s sample.pytest.shell`, we have a wrapper and some other magic in there but hey it works, did not realize that there was a pytest sample. Thanks again.

gchwier commented:
>  we have a wrapper and some other magic in there but hey it works, did not realize that there was a pytest sample. Thanks again.

You can find some details here:
https://docs.zephyrproject.org/latest/develop/test/pytest.html
Now there are also three other harnesses, that uses pytest:
'harness: shell', 'harness: power', 'harness: display_capture'
More tests are in downstream projects probable, so good to add new options like `--flash-command` also in pytest plugin

marc-hb reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/pull/93944#discussion_r2279292600

> tl;dr, do things like this work fine? They should, just making sure.

Pong.

fabiobaltieri commented:
- tweaked the harness commit to do the parsing in the same way as the twister flag, i.e. pass an array, not sure why I did not do it in like that the first time
- added a patch to switch the extra arguments parsing from split to csv, that handles correctly quotes and allows passing comma arguments, which sounds like a not so unreasonable use case

marc-hb reviewed (commented):
> twister: use csv to split extra argument arguments

I'm sorry to ask this but I think this deserves a separate, preliminary PR for both 1) visibility and 2) CI reasons.

\2) give upstream and downstream CIs at least a 24h opportunity for one "nightly" build to catch any test failure caused by this. This change is for the greater good but it is technically backwards-incompatible. Maybe some tests already had some "creative" quoting solution that is obviously going to be incompatible with this.

\1) Give people an opportunity to notice this change in their flood of GitHub notifications before it gets merged. And make it easily to spot even after merge when scanning what changed (and possibly broke some test of theirs)

Should be really quick
https://github.com/zephyrproject-rtos/zephyr/pull/83839/files#diff-d84ff1b1f7f14c50055618e1e0f6a0986cae639ab168881fc493de64ebd76971

fabiobaltieri commented:
@marc-hb cool, dropped the last commit, let's merge the others once they are in I'll open a separate PR for the quotes magic stuff.

marc-hb commented:
Respectfully but strongly disagree with the "process" question here.

1) I think it's unlikely, but if the backwards-incompatible quoting fix needs to be reverted, then we have IMHO discovered a "bigger" and more important problem in the existing codebase, bigger and more urgent than adding this new feature that is not there yet and has been under review for almost 2 months. The former should be dealt with before the latter. 

> whoever does the rollback may (rightfully so) not care enough to figure what's going on and end up reverting both.

2) If there's at least 24h distance between the two things and they don't git conflict with each other, AND there is a regression with the first thing AND it does not get caught in the first 24h... then there is still zero chance of both being reverted at the same time by accident. Reverts get code reviews too.

marc-hb commented:
> then there is still zero chance of both being reverted at the same time by accident. Reverts get code reviews too.

BTW people sometimes get emotional about reverts but I think a small number of infrequent reverts is perfectly fine. 

"If you never break anything, then you're not trying hard enough" (a former team lead).

fabiobaltieri commented:
Or I can add it back to the stack of the PR, you tell me.

marc-hb commented:
The "urgent" problem is not fixing the quoting bug itself; I understand it's been there since forever. The urgent questions are: 1) finding whether anything depends on the current behavior 2) if anything, figuring out what to do about it/them. Can they be updated easily? Or is that too much branching hassle? Other unexpected problem?

Knowing where we stand and what the next steps are is IMHO higher priority than adding a brand new feature interacting with that very same, likely changing behavior.

But again: in all likelihood this small and isolated quoting fix will just sail through and not break anything. So I don't understand the rush (after 2 months) to get this new feature first.

fabiobaltieri commented:
Done, quote handled properly, no chances in behavior for any current flag.

marc-hb reviewed (commented):
1st commit LGTM. I don't know the test harness.

fabiobaltieri commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -544,7 +545,19 @@ def run_custom_script(script, timeout):
 
+    def _create_flash_command(self, hardware):
+        command = next(csv.reader([self.options.flash_command]))
+        command.extend(['--build-dir', self.build_dir])
```
Not sure I get the comment, you want me to put `--build-dir` and `--board-id` before the extended flags? so it'd be `command[0]` then the flags then any eventual other user flag?

marc-hb replied:
Thanks! However I think you forgot to adjust the example accordingly?

fabiobaltieri commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -836,6 +836,15 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
     )
+    parser.add_argument(
+        "--flash-command",
+        help="""Uses a custom flash command to flash when running with
```
done

nashif commented:
https://github.com/zephyrproject-rtos/zephyr/pull/95298 should fix those test failures

fabiobaltieri commented:
Thanks, looks good now, had to rebase as well, looks like the workflow does not do it.

@marc-hb @gchwier @hakehuang feel free to take another look, should have addressed everything, sorry for the spotty updates, was out at ZDS.

marc-hb commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -1130,23 +1141,39 @@ def mock_availability(handler, instance, no=num_of_failures):
+        None,
+        'path to/flash_command,with,args,"1,2,3",4',
+        ['path to/flash_command', '--build-dir', '$build_dir', '--board-id',
+         12345, 'with', 'args', '1,2,3', '4']
```
How about some space in the arg too? I mean '1, 2, 3' for instance.

marc-hb commented:
Fixups submitted in
- #95442
